datalad_crawler/pipelines/loris.py

```diff
@@ -0,0 +1,90 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""A pipeline for crawling a LORIS database"""
+
+# Import necessary nodes
+from ..nodes.crawl_url import crawl_url
+from datalad.utils import updated
+from ..nodes.annex import Annexificator
+from datalad_crawler.consts import DATALAD_SPECIAL_REMOTE
+from os.path import basename
+
+
+import json
+# Possibly instantiate a logger if you would like to log
+# during pipeline creation
+from logging import getLogger
+lgr = getLogger("datalad.crawler.pipelines.kaggle")
+
+
+class LorisAPIExtractor(object):
+
+    def __init__(self, apibase=None, annex=None):
+        self.apibase = apibase
+        self.meta    = {}
+        self.repo    = annex.repo
+
+    def __call__(self, data):
+        jsdata = json.loads(data["response"])
+        for candidate in jsdata["Images"]:
+            filename = basename(candidate["Link"])
+            self.meta[filename] = candidate
+            yield updated(data, { "url" : self.apibase + candidate["Link"] })
+        return
+
+    def finalize(self):
+        def _finalize(data):
+            for filename in self.meta:
+                metadata_setter = self.repo.set_metadata(
```

Comment:
I would really like to see a good example of a dataset with metadata. Git annex level to store metadata is convenient but it's limitations, such as metadata is assigned per annex key and but filename, could be limiting factor

Reply:
What kind of metadata would you like to see? For now, this is just annotating the data that comes along with the file in the API when getting the list. (I want to expand this to add other metadata from the API which is available at other endpoints, but it should be easy to add that incrementally).

The annex level seemed easier because:
1. The extractor is in the base datalad install
2. It's arbitrary JSON returned by a REST API, so there's no conversion necessary to some other format if we just treat it as annex level key/value pairs

Reply:
after talking about this yesterday: the plan is to store the metadata in a JSON file alongside the data and write a separate metadata extractor in datalad-neuroimaging to handle extracting it.